### PR TITLE
PXP-568: [CLI] Make the application instances connect interactive

### PR DESCRIPTION
**Pull request by mfauzaan**
## Summary

Make the instances connect command interactive, similarly with the deployment execute command, it will be a multiple steps session, like the example below

Step 1:
Screenshot 2023-07-11 at 3 32 56 PM

Step 2:
Screenshot 2023-07-11 at 3 33 05 PM

Step 3:
Screenshot 2023-07-11 at 3 33 14 PM

Ticket: [PXP-568](https://mindvalley.atlassian.net/browse/PXP-568)

## What's Changed

[PXP-568]: https://mindvalley.atlassian.net/browse/PXP-568?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/118#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: src/commands/application/instances/connect.rs
Comment (jk-gan) [inline src/commands/application/instances/connect.rs:70]:
```suggestion
    if !has_permission(&client, &application, &namespace, &version)
        .await?
```

I totally missed this, this should be handled.

File: src/commands/application/instances/connect.rs
Comment (jk-gan) [inline src/commands/application/instances/connect.rs:37]: I think before we should the selection, we should show the current application first:
```
Current application: `wukong-ci-mock`.

? Please choose the namespace you want to connect to ›
❯ Prod
  Staging 
```

(You can refer to `wukong deployment execute`

File: src/commands/application/instances/mod.rs
Comment (jk-gan) [inline src/commands/application/instances/mod.rs:34]: @onimsha Do we want to provide some optional flags similar to `wukong deployment execute --namespace pro --version green`?

**Review by jk-gan — CHANGES_REQUESTED** [review]

**Comment by jk-gan** [comment]
Btw, make sure to update the completions using `cargo build --release && ./generate-completions.sh`.

**Comment by jk-gan** [comment]
maybe it is just me but the `[1/4]` and `[2/4]` doesn't make sense at here because the steps are split with 2 questions
Screenshot 2023-07-12 at 1 40 00 PM

**Comment by onimsha** [comment]
@mfauzaan we need to handle this error as well, since some application won't have certain `namespace` or `version`

**Case 1** : Application doesn't have any k8s_cluster config for `blue` version. 

```bash
cargo run -- application instances connect --application mv-wukong-ci-mock
    Finished dev [unoptimized + debuginfo] target(s) in 1.02s
     Running `target/debug/wukong application instances connect --application mv-wukong-ci-mock`
✔ Please choose the namespace you want to connect to · Staging
✔ Please choose the version you want to connect to · Blue
[1/4] ⠐ Checking your permission to connect to the remote instance...           
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: APIError(ResponseError { code:
"k8s_cluster_version_config_not_defined", message: "isAuthorized:6:3: k8s_cluster_version_config_not_defined" })', /Users/tuanco/code-workspaces/rust/src/github.com/mindvalley/wukong-cli/src/commands/application/instances/connect.rs:70:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Case 2** : Application does not have any k8s_cluster in any of the version. 
```bash
❯ cargo run -- application instances connect --application mv-submgr
    Finished dev [unoptimized + debuginfo] target(s) in 1.53s
     Running `target/debug/wukong application instances connect --application mv-submgr`
✔ Please choose the namespace you want to connect to · Staging
✔ Please choose the version you want to connect to · Green
Error - API Response Error: isAuthorized:6:3: k8s_cluster_namespace_config_not_defined

```

In either case, we should display the error accordance to our current error handling logic, which shows the error and the suggestion. @jk-gan may give you the instruction about how to format the error and the suggestion in the CLI outputs.

File: src/commands/application/instances/mod.rs
Comment (onimsha) [reply src/commands/application/instances/mod.rs:34]: @jk-gan I think we should, since it will aligns with what we're supporting in the interactive sessions of other commands as well.

**Comment by mfauzaan** [comment]
> @mfauzaan we need to handle this error as well, since some application won't have certain `namespace` or `version`
> 
> **Case 1** : Application doesn't have any k8s_cluster config for `blue` version.
> 
> ```shell
> cargo run -- application instances connect --application mv-wukong-ci-mock
>     Finished dev [unoptimized + debuginfo] target(s) in 1.02s
>      Running `target/debug/wukong application instances connect --application mv-wukong-ci-mock`
> ✔ Please choose the namespace you want to connect to · Staging
> ✔ Please choose the version you want to connect to · Blue
> [1/4] ⠐ Checking your permission to connect to the remote instance...           
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: APIError(ResponseError { code:
> "k8s_cluster_version_config_not_defined", message: "isAuthorized:6:3: k8s_cluster_version_config_not_defined" })', /Users/tuanco/code-workspaces/rust/src/github.com/mindvalley/wukong-cli/src/commands/application/instances/connect.rs:70:10
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> **Case 2** : Application does not have any k8s_cluster in any of the version.
> 
> ```shell
> ❯ cargo run -- application instances connect --application mv-submgr
>     Finished dev [unoptimized + debuginfo] target(s) in 1.53s
>      Running `target/debug/wukong application instances connect --application mv-submgr`
> ✔ Please choose the namespace you want to connect to · Staging
> ✔ Please choose the version you want to connect to · Green
> Error - API Response Error: isAuthorized:6:3: k8s_cluster_namespace_config_not_defined
> ```
> 
> In either case, we should display the error accordance to our current error handling logic, which shows the error and the suggestion. @jk-gan may give you the instruction about how to format the error and the suggestion in the CLI outputs.

Added the error messages by using existing error codes used in deployment pipeline, let me know if we need to add suggestions which does not have any on pipeline commands as well

Screenshot 2023-07-13 at 2 38 07 PM

File: src/commands/application/instances/mod.rs
Comment (mfauzaan) [reply src/commands/application/instances/mod.rs:34]: So basically, user can skip the selection by passing as a optional parameter for all including pod name?

Comment (onimsha) [reply src/commands/application/instances/mod.rs:34]: no, the pod name is not a parameter, only the `--namspace` and `--version`. In the event that any of these parameters are set, the CLI will skip a specific step, or both of the steps of selecting `namespace` and `version`.

Only the last step, which is selecting the pod, is mandatory in the interactive session.

Comment (jk-gan) [reply src/commands/application/instances/mod.rs:34]: what about we still provide the current method to connect? 
For example, 
- `wukong application instances connect green@staging/mv-xxx-yyy` -> we will use the existing flow
- `wukong application instances connect --namespace staging --version green` -> we show the interactive flow, but skip the step when the info is provided from the flags
- `wukong application instances connect` -> we show the full interactive flow

Comment (mfauzaan) [reply src/commands/application/instances/mod.rs:34]: Wouldn't it be confusing to the user?

**Comment by onimsha** [comment]
> > @mfauzaan we need to handle this error as well, since some application won't have certain `namespace` or `version`
> > **Case 1** : Application doesn't have any k8s_cluster config for `blue` version.
> > ```shell
> > cargo run -- application instances connect --application mv-wukong-ci-mock
> >     Finished dev [unoptimized + debuginfo] target(s) in 1.02s
> >      Running `target/debug/wukong application instances connect --application mv-wukong-ci-mock`
> > ✔ Please choose the namespace you want to connect to · Staging
> > ✔ Please choose the version you want to connect to · Blue
> > [1/4] ⠐ Checking your permission to connect to the remote instance...           
> > thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: APIError(ResponseError { code:
> > "k8s_cluster_version_config_not_defined", message: "isAuthorized:6:3: k8s_cluster_version_config_not_defined" })', /Users/tuanco/code-workspaces/rust/src/github.com/mindvalley/wukong-cli/src/commands/application/instances/connect.rs:70:10
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Case 2** : Application does not have any k8s_cluster in any of the version.
> > ```shell
> > ❯ cargo run -- application instances connect --application mv-submgr
> >     Finished dev [unoptimized + debuginfo] target(s) in 1.53s
> >      Running `target/debug/wukong application instances connect --application mv-submgr`
> > ✔ Please choose the namespace you want to connect to · Staging
> > ✔ Please choose the version you want to connect to · Green
> > Error - API Response Error: isAuthorized:6:3: k8s_cluster_namespace_config_not_defined
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In either case, we should display the error accordance to our current error handling logic, which shows the error and the suggestion. @jk-gan may give you the instruction about how to format the error and the suggestion in the CLI outputs.
> 
> Added the error messages by using existing error codes used in deployment pipeline, let me know if we need to add suggestions which does not have any on pipeline commands as well @onimsha
> 
> Screenshot 2023-07-13 at 2 38 07 PM

@mfauzaan I think we can't re-use the existing error messages for these 2:

For the first error, `k8s_cluster_version_config_not_defined`, we should display this 
```
Error: The {version} version has no associated k8s cluster configuration.
Suggestion: You can try to check the following:
                      * Whether your application is supporting this {version} version. 
                      * Contact Wukong dev team to check if there is any k8s configuration enabled for this version. 
```

For the second error, `k8s_cluster_namespace_config_not_defined`, we should display this
```
Error: There is no k8s configuration associated with your application. 
Suggestion: You may want to contact Wukong dev team to check if there is any k8s configuration for your application.
```

File: src/commands/application/instances/mod.rs
Comment (jk-gan) [reply src/commands/application/instances/mod.rs:34]: ya, actually quite confusing haha

File: src/error.rs
Comment (onimsha) [inline src/error.rs:132]:
```suggestion
    #[error("This version has no associated k8s cluster configuration.")]
```

File: src/error.rs
Comment (onimsha) [inline src/error.rs:138]:
```suggestion
    #[error("This application is not available  in k8s.")]
```

**Review by onimsha — CHANGES_REQUESTED** [review]

File: src/commands/application/instances/mod.rs
Comment (jk-gan) [inline src/commands/application/instances/mod.rs:37]: This should be `enum` instead of `String`, it will show the possible values in `--help` as well as show nice error message if user input wrong value
Screenshot 2023-07-14 at 12 29 45 PM
Screenshot 2023-07-14 at 12 30 43 PM

File: src/commands/application/instances/mod.rs
Comment (jk-gan) [inline src/commands/application/instances/mod.rs:41]: this is the same as above

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/error.rs
Comment (jk-gan) [reply src/error.rs]: I think we should show the application name, either here or in the suggestion, so the user can report the issue easily by just show this message, then the PIC will know which application is not available.

**Comment by jk-gan** [comment]
Screenshot 2023-07-14 at 1 22 07 PM
The UX for this is not great, we should inform user when they've selected even they are using the flags. For example, you can refer to `wukong deployment execute --namespace prod`:
Screenshot 2023-07-14 at 1 22 23 PM

File: src/commands/application/instances/connect.rs
Comment (jk-gan) [inline src/commands/application/instances/connect.rs:398]: Question, why the return value need to be `Option`? Is it possible to get the index outside of the selection?

File: src/commands/application/instances/connect.rs
Comment (jk-gan) [inline src/commands/application/instances/connect.rs:420]: same here, why this need to be Option?

**Review by jk-gan — APPROVED** [review]
Overall, LGTM

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/118#pullrequestreview-1529727175" title="Approved">jk-gan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/118#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: src/commands/application/instances/connect.rs
Comment (mfauzaan) [reply src/commands/application/instances/connect.rs:398]: Its not because of that but when I moved the selection to a function its expecting to have with default state which is null when user uses the ctr+c

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/118#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/118#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by onimsha — APPROVED** [review]

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/118#" title="Author self-approved">mfauzaan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/118#pullrequestreview-1529926271" title="Approved">onimsha</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
